examples/gno.land/p/sys/vals/pos/pos_test.gno

```diff
@@ -0,0 +1,312 @@
+		)
+
+		// Create the protocol with no initial set
+		p := NewPoS()
+
+		// Set the origin caller
+		std.TestSetOrigCaller(callerAddress)
+		std.TestSetOrigSend(value, std.Coins{})
+
+		for i := int64(0); i < numTries; i++ {
+			if p.IsValidator(callerAddress) {
+				t.Fatalf("should not have a validator at try %d", i)
+			}
+
+			// Attempt to add the validator
+			testing.NotPanics(t, func() {
+				p.AddValidator(callerAddress, callerKey)
+			})
+		}
+
+		// Make sure the user became a validator
+		if !p.IsValidator(callerAddress) {
+			t.Fatalf("should be a validator")
+		}
+	})
+
+	t.Run("user becomes validator in single stake (threshold)", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			callerAddress = std.Address("caller")
+			callerKey     = "public-key"
+
+			value = std.Coins{{"ugnot", stakeThreshold.Amount}}
+		)
+
+		// Create the protocol with an initial set
+		p := NewPoS()
+
+		// Set the origin caller
+		std.TestSetOrigCaller(callerAddress)
+		std.TestSetOrigSend(value, std.Coins{})
+
+		// Make sure the caller is not a validator
+		if p.IsValidator(callerAddress) {
+			t.Fatalf("should not be a validator")
+		}
+
+		// Attempt to add the validator
+		testing.NotPanics(t, func() {
+			p.AddValidator(callerAddress, callerKey)
+		})
+
+		// Make sure the user became a validator
+		if !p.IsValidator(callerAddress) {
+			t.Fatalf("should be a validator")
+		}
+	})
+}
+
+func TestPoS_IsValidator(t *testing.T) {
+	t.Parallel()
+
+	t.Run("address is not in the set", func(t *testing.T) {
+		t.Parallel()
+
+		// Empty PoS set
+		p := NewPoS()
+
+		if p.IsValidator(std.Address("random address")) {
+			t.Fatal("should not be a validator")
+		}
+	})
+
+	t.Run("address is in the set", func(t *testing.T) {
+		t.Parallel()
+
+		initialSet := generateTestValidators(10)
+
+		// Existing PoS set
+		p := NewPoS(WithInitialSet(initialSet))
+
+		if !p.IsValidator(initialSet[0].Address) {
+			t.Fatal("should be a validator")
+		}
+	})
+}
+
+func TestPoS_GetValidators(t *testing.T) {
+	t.Parallel()
+
+	t.Run("empty set", func(t *testing.T) {
+		t.Parallel()
+
+		// Empty PoS set
+		p := NewPoS()
+
+		if len(p.GetValidators()) != 0 {
+			t.Fatal("should be empty set")
+		}
+	})
+
+	t.Run("existing set", func(t *testing.T) {
+		t.Parallel()
+
+		initialSet := generateTestValidators(10)
+
+		// Existing PoS set
+		p := NewPoS(WithInitialSet(initialSet))
+
+		vals := p.GetValidators()
+
+		// Make sure the set has a valid length
+		if len(vals) != len(initialSet) {
+			t.Fatal("invalid set length")
+		}
+
+		// Make sure each validator is valid
+		for i, val := range vals {
+			ref := initialSet[i]
+
+			if val.Address != ref.Address {
+				t.Fatal("address mismatch")
+			}
+
+			if val.PubKey != ref.PubKey {
+				t.Fatal("address mismatch")
+			}
+
+			if val.VotingPower != 0 {
+				t.Fatal("invalid voting power")
+			}
+		}
+	})
+}
+
+func TestPoS_RemoveValidator_Invalid(t *testing.T) {
+	t.Parallel()
```

Comment:
would love to see this disappear thanks to #2376 :) 🤞

Reply:
Oooh, this is interesting 👀 

I'm curious if there is a strong case to move away from Go's explicit parallel